cmd/runtimetest/main.go

```diff
@@ -609,21 +609,71 @@ func validateMountsExist(spec *rspec.Spec) error {
 		mountsMap[mountInfo.Mountpoint] = append(mountsMap[mountInfo.Mountpoint], m)
 	}
 
-	for _, specMount := range spec.Mounts {
-		if specMount.Type == "bind" || specMount.Type == "rbind" {
+	for _, configMount := range spec.Mounts {
+		if configMount.Type == "bind" || configMount.Type == "rbind" {
 			// TODO: add bind or rbind check.
 			continue
 		}
 
 		found := false
-		for _, sysMount := range mountsMap[filepath.Clean(specMount.Destination)] {
-			if err := mountMatch(specMount, sysMount); err == nil {
+		for _, sysMount := range mountsMap[filepath.Clean(configMount.Destination)] {
+			if err := mountMatch(configMount, sysMount); err == nil {
 				found = true
 				break
 			}
 		}
 		if !found {
-			return fmt.Errorf("Expected mount %v does not exist", specMount)
+			return fmt.Errorf("Expected mount %v does not exist", configMount)
+		}
+	}
+
+	return nil
+}
+
+func validateMountsOrder(spec *rspec.Spec) error {
+	if runtime.GOOS == "windows" {
+		logrus.Warnf("mounts order validation not yet implemented for OS %q", runtime.GOOS)
+		return nil
+	}
+
+	mountInfos, err := mount.GetMounts()
+	if err != nil {
+		return err
+	}
+
+	type mountOrder struct {
+		Order  int
+		Root   string
+		Dest   string
+		Source string
+	}
+	mountsMap := make(map[string][]mountOrder)
+	for i, mountInfo := range mountInfos {
+		m := mountOrder{
+			Order:  i,
+			Root:   mountInfo.Root,
+			Dest:   mountInfo.Mountpoint,
+			Source: mountInfo.Source,
+		}
+		mountsMap[mountInfo.Mountpoint] = append(mountsMap[mountInfo.Mountpoint], m)
+	}
+	current := -1
+	for _, configMount := range spec.Mounts {
+		mounts := mountsMap[configMount.Destination]
+		for _, mount := range mounts {
+			source := mount.Source
+			for _, option := range configMount.Options {
+				if option == "bind" || option == "rbind" {
+					source = mount.Root
+				}
+			}
+			if source == configMount.Source {
+				if current > mount.Order {
+					return fmt.Errorf("%s is not mounted in order", configMount.Source)
```

Comment:
> Can you give an example?

I haven't compiled the following, but something like:

```go
var current *mountOrder
for i, configMount := range spec.Mounts {
  mounts, ok := mountsMap[configMount.Destination]
  if !ok {
    // this error message could be better too, but I'm following validateMountsExist
    return fmt.Errorf("Expected mount %v does not exist", configMount)
  }
  for j, mount := range mounts {
    source := mount.Source
    for _, option := range configMount.Options {
      if option == "bind" || option == "rbind" {
        source = mount.Root
        break
      }
    }
    if source == configMount.Source {
      if current != nil && current.Order > mount.Order {
        return fmt.Errorf("mounts[%d] %q is mounted before %q", i, configMount.Destination, current.Destination)
      }
      current = &mount
      // in order to deal with dup mount elements
      mountsMap[configMount.Destination] = append(mountsMap[configMount.Destination][:j], mountsMap[configMount.Destination][j+1:]...)
      break
    }
  }
}
```

I'm also not clear on your dup-handling bit at the end.  And we'd probably have better (or at least more consistent) matching if we used `mountMatch` here too (loke we already do [here][1]).

[1]: https://github.com/opencontainers/runtime-tools/blob/9e0e42dbf918070406a2a4a2e1476e7350ba9129/cmd/runtimetest/main.go#L619-L627